alacritty_terminal/src/renderer/mod.rs

```diff
@@ -1002,74 +1067,138 @@ impl<'a> RenderApi<'a> {
             // zero-width characters inside the preceding character, the
             // anchor has been moved to the right by one cell.
-            glyph.left += glyph_cache.metrics.average_advance as f32;
+            glyph.left += average_advance;
 
             self.add_render_item(&cell, &glyph);
         }
     }
+
+    pub fn render_text_run(&mut self, text_run: TextRun, glyph_cache: &mut GlyphCache) {
+        match &text_run.content {
+            TextRunContent::Cursor(cursor_key) => {
+                self.render_cursor(&text_run.start_cell(), *cursor_key, glyph_cache)
+            },
+            TextRunContent::CharRun(run, zero_widths) => {
+                // Get font key for cell
+                let font_key = Self::determine_font_key(text_run.flags, glyph_cache);
+
+                let shaped_glyphs = if text_run.flags.contains(cell::Flags::HIDDEN) {
+                    GlyphIter::Hidden(std::iter::repeat(hidden_glyph(font_key, glyph_cache, self)))
+                } else {
+                    GlyphIter::Shaped(glyph_cache.shape_run(&run, font_key, self).into_iter())
+                };
+
+                for ((cell, glyph), zero_width_chars) in
+                    text_run.cells().zip(shaped_glyphs).zip(zero_widths.iter())
+                {
+                    self.add_render_item(&cell, &glyph);
+                    if text_run.flags.contains(cell::Flags::WIDE_CHAR) {
+                        // Manually add instance data for WIDE_CHAR_SPACER
+                        // Renderer only renders single character width rectangles
+                        // If we don't add in this instance data then we only render one cell of
+                        // fg/bg/strikethrough/etc. for WIDE_CHARs
+                        let glyph = hidden_glyph(font_key, glyph_cache, self);
+                        self.add_render_item(
+                            &RenderableCell { column: cell.column + 1, ..cell.clone() },
+                            &glyph,
+                        );
+                    }
+                    self.render_zero_widths(
+                        zero_width_chars.iter().filter(|c| **c != ' '),
+                        &cell,
+                        font_key,
+                        glyph_cache,
+                    );
+                }
+            },
+        };
+    }
+}
+
+/// Returns an infinite iterator of hidden glyphs for a given font key and size.
+fn hidden_glyph<'a, L>(font_key: FontKey, glyph_cache: &'a mut GlyphCache, loader: &mut L) -> Glyph
+where
+    L: LoadGlyph,
+{
+    // If text_run is hidden we don't want to shape text.
+    // But we still want to run each cell through add_render_item so colors get handled
+    // appropiately. We construct a dummy key that should be cached a majority of the time so
+    // that we have a valid glyph.
+    let key = GlyphKey { id: PLACEHOLDER_GLYPH, font_key, size: glyph_cache.font_size };
+    *glyph_cache.get(key, loader)
+}
```

Comment:
Does a Default glyph behave correctly for any of the font key/size combos? My concern with this is that since it doesn't account for font size it's spacing may not relative too large or small font sizes. However it's also possible all the spacing stuff is handled by the RenderableCell that gets passed and the glyph layout members aren't used at all for that purpose.

Reply:
Based on my testing, everything should work perfectly. The only issue I saw was that the default texture ID might not be defined, but since we always start with texture id 0 and number types default to 0, we should always have the texture id 0 valid.

Feel free to test this yourself though.